modules/storages/spec/features/view_project_storage_members_spec.rb

```diff
@@ -0,0 +1,121 @@
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+require_relative '../spec_helper'
+
+RSpec.describe(
+  'Project storage members connection status view',
+  with_flag: { storage_project_members_check: true }
+) do
+  let(:user) { create(:user) }
+  let(:admin_user) { create(:admin) }
+  let(:connected_user) { create(:user) }
+  let(:connected_no_permissions_user) { create(:user) }
+  let(:disconnected_user) { create(:user) }
+
+  let!(:storage) { create_nextcloud_storage_with_oauth_application }
+  let!(:project) { create_project_with_storage_and_members }
+  let!(:project_storage) { create(:project_storage, project:, storage:) }
+  let(:oauth_client) { create(:oauth_client, integration: storage) }
+
+  before do
+    create_oauth_client_tokens_for_users(oauth_client:,
+                                         users: [connected_user, admin_user,
+                                                 connected_no_permissions_user])
+  end
+
+  it 'lists project members connection statuses' do
+    login_as user
+
+    # Go to Projects -> Settings -> File Storages
+    visit project_settings_project_storages_path(project)
+
+    expect(page).to have_title('File storages')
+    expect(page).to have_text(storage.name)
+    page.find('.icon.icon-group').click
+
+    # Members check page
+    expect(page).to have_current_path project_settings_project_storage_members_path(project_id: project,
+                                                                                    project_storage_id: project_storage)
+
+    [
+      [user, 'Not connected'],
+      [admin_user, 'Connected'],
+      [connected_user, 'Connected'],
+      [connected_no_permissions_user, 'User role has no storages permissions'],
+      [disconnected_user, 'Not connected']
+    ].each do |(principal, status)|
+      expect(page).to have_selector("#member-#{principal.id} .name", text: principal.name)
+      expect(page).to have_selector("#member-#{principal.id} .status", text: status)
+    end
```

Comment:
Something I like to do (which helps with debugging if ever need be) is to wrap these kind of "bulk expectations" in an [aggregate_failures](https://rubydoc.info/github/rspec/rspec-expectations/RSpec%2FMatchers:aggregate_failures) block. That way you get to see - out of all the expected connection statuses - which didn't match the expectations, instead of having to run the spec multiple times to find out consecutively that precisely `admin_user`, `connected_user` and connected_no_permissions_user` are the expectations failing. You get to find out immediately in a single spec run what is off.

```suggestion
    aggregate_failures 'Verifying Connection Statuses' do
      [
        [user, 'Not connected'],
        [admin_user, 'Connected'],
        [connected_user, 'Connected'],
        [connected_no_permissions_user, 'User role has no storages permissions'],
        [disconnected_user, 'Not connected']
      ].each do |(principal, status)|
        expect(page).to have_selector("#member-#{principal.id} .name", text: principal.name)
        expect(page).to have_selector("#member-#{principal.id} .status", text: status)
      end
    end
```

Reply:
Sweet, neat application! 🎨

Reply:
Super. Nice one @aaron-contreras 👍🏾 

```ruby
Failures:

  1) Project storage members connection status view lists project members connection statuses
     Got 5 failures from failure aggregation block "Verifying Connection Statuses".
     # ./modules/storages/spec/features/view_project_storage_members_spec.rb:66:in `block (2 levels) in <top (required)>'
     # ./spec/support/capybara.rb:16:in `block (2 levels) in <top (required)>'
     # /usr/local/bundle/gems/webmock-3.18.1/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'
     # ./spec/support/shared/with_env.rb:58:in `block (2 levels) in <top (required)>'
     # ./spec/support/shared/with_direct_uploads.rb:199:in `block (2 levels) in <top (required)>'
     # /usr/local/bundle/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:124:in `block in run'
     # /usr/local/bundle/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `loop'
     # /usr/local/bundle/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `run'
     # /usr/local/bundle/gems/rspec-retry-0.6.2/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
     # /usr/local/bundle/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:37:in `block (2 levels) in setup'

     1.1) Failure/Error: expect(page).to have_selector("#member-#{principal.id} .status", text: status)
            expected to find visible css "#member-195 .status" with text "Not connectedd" but there were no matches. Also found "Not connected", which matched the selector but not all filters.
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:75:in `block (4 levels) in <top (required)>'
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:73:in `each'
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:73:in `block (3 levels) in <top (required)>'

     1.2) Failure/Error: expect(page).to have_selector("#member-#{principal.id} .status", text: status)
            expected to find visible css "#member-196 .status" with text "Connectedd" but there were no matches. Also found "Connected", which matched the selector but not all filters.
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:75:in `block (4 levels) in <top (required)>'
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:73:in `each'
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:73:in `block (3 levels) in <top (required)>'

     1.3) Failure/Error: expect(page).to have_selector("#member-#{principal.id} .status", text: status)
            expected to find visible css "#member-197 .status" with text "Connectedd" but there were no matches. Also found "Connected", which matched the selector but not all filters.
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:75:in `block (4 levels) in <top (required)>'
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:73:in `each'
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:73:in `block (3 levels) in <top (required)>'

     1.4) Failure/Error: expect(page).to have_selector("#member-#{principal.id} .status", text: status)
            expected to find visible css "#member-198 .status" with text "User role has no storages permissionsd" but there were no matches. Also found "User role has no storages permissions", which matched the selector but not all filters.
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:75:in `block (4 levels) in <top (required)>'
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:73:in `each'
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:73:in `block (3 levels) in <top (required)>'

     1.5) Failure/Error: expect(page).to have_selector("#member-#{principal.id} .status", text: status)
            expected to find visible css "#member-199 .status" with text "Not connectedd" but there were no matches. Also found "Not connected", which matched the selector but not all filters.
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:75:in `block (4 levels) in <top (required)>'
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:73:in `each'
          # ./modules/storages/spec/features/view_project_storage_members_spec.rb:73:in `block (3 levels) in <top (required)>'
```